### PR TITLE
CLN refactor rest of core

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Some refactorings found by Sourcery https://sourcery.ai/

I've removed the ones of the kind
```diff
- if param:
-     var = a
- else:
-     var = b
+ var = a if param else b
```

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -5410,19 +5405,15 @@ def __setattr__(self, name: str, value) -> None:
...
             try:
                 existing = getattr(self, name)
-                if isinstance(existing, Index):
+                if isinstance(existing, Index) or name not in self._info_axis:
                     object.__setattr__(self, name, value)
-                elif name in self._info_axis:
-                    self[name] = value
                 else:
-                    object.__setattr__(self, name, value)
+                    self[name] = value
```
Comment (jreback) [inline pandas/core/generic.py:5416]: can you check the logic that was changed here (IIRC this is for performance but the change might be good)

**Review by jreback — CHANGES_REQUESTED** [review]
one comment otherwise lgtm

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -100,11 +100,13 @@ def _ensure_data(
...
 
     # we check some simple dtypes first
-    if is_object_dtype(dtype):
-        return ensure_object(np.asarray(values)), np.dtype("object")
-    elif is_object_dtype(values) and dtype is None:
+    if (
+        is_object_dtype(dtype)
+        or not is_object_dtype(dtype)
+        and is_object_dtype(values)
+        and dtype is None
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:107]: can you put parens where appropriate to make this more obvious

File: pandas/core/base.py [hunk 3]
```diff
@@ -1244,8 +1242,7 @@ def searchsorted(self, value, side="left", sorter=None) -> np.ndarray:
 
     def drop_duplicates(self, keep="first"):
         duplicated = self.duplicated(keep=keep)
-        result = self[np.logical_not(duplicated)]
-        return result
+        return self[np.logical_not(duplicated)]
```
Comment (jbrockmendel) [inline pandas/core/base.py:1245]: id prefer `~duplicated` over logical_not

File: pandas/core/generic.py [hunk 4]
```diff
@@ -8857,16 +8851,18 @@ def _align_series(
...
+            and is_datetime64tz_dtype(left.index.dtype)
+            and left.index.tz != right.index.tz
+            and join_index is not None
+        ):
+            # GH#33671 ensure we don't change the index on
+            #  our original Series (NB: by default deep=False)
+            left = left.copy()
+            right = right.copy()
+            left.index = join_index
+            right.index = join_index
```
Comment (jbrockmendel) [inline pandas/core/generic.py:8865]: identical(?) to chunk above, could make a helper.  also im pretty sure mypy is going to complain about Index not having a tz attribute

File: pandas/core/nanops.py [hunk 5]
```diff
@@ -185,14 +182,11 @@ def _get_fill_value(
             else:
                 return -np.inf
     else:
-        if fill_value_typ is None:
-            return iNaT
+        if fill_value_typ is not None and fill_value_typ == "+inf":
```
Comment (jbrockmendel) [inline pandas/core/nanops.py:185]: i think can just do the str comparison?

File: pandas/core/nanops.py [hunk 6]
```diff
@@ -395,15 +389,12 @@ def _na_for_min_count(
     if fill_value is NaT:
         fill_value = values.dtype.type("NaT", "ns")
 
-    if values.ndim == 1:
-        return fill_value
-    elif axis is None:
+    if values.ndim == 1 or axis is None:
```
Comment (jbrockmendel) [inline pandas/core/nanops.py:392]: i think this is a case where its useful to see we get coverage of both cases

File: pandas/core/sorting.py [hunk 7]
```diff
@@ -218,7 +218,7 @@ def decons_group_index(comp_labels, shape):
     x = comp_labels
     for i in reversed(range(len(shape))):
         labels = (x - y) % (factor * shape[i]) // factor
-        np.putmask(labels, comp_labels < 0, -1)
+        np.putmask(labels, x < 0, -1)
```
Comment (jbrockmendel) [inline pandas/core/sorting.py:221]: lets just do `labels[x < 0] = -1`, putmask has different semantics when the last arg is a listlike but thats not relevant here

File: pandas/core/generic.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/core/generic.py:8838]: @jbrockmendel do you know when this code is meant to be reached? Running
```
pytest pandas/tests/generic/
```
doesn't cover it. I couldn't get here with `test_fast.sh` either, even though [codecov shows it as covered](https://codecov.io/gh/pandas-dev/pandas/src/master/pandas/core/generic.py)

Comment (jbrockmendel) [reply pandas/core/generic.py:8838]: Looks like `test_alignment_doesnt_change_tz` from #36503 hits the other usage of this pattern, might not have gotten both

**Comment by jreback** [comment]
can you merge master ping on green

**Comment by MarcoGorelli** [comment]
I still need to address https://github.com/pandas-dev/pandas/pull/37586#discussion_r519926657 , will ping when it's ready

**Comment by jreback** [comment]
moving off 1.2 until fixed up

**Comment by MarcoGorelli** [comment]
Sure, no problem, though I believe I've responded to the review comments, I think it should be ready

File: pandas/core/sorting.py [hunk 8]
```diff
@@ -218,7 +218,7 @@ def decons_group_index(comp_labels, shape):
     x = comp_labels
     for i in reversed(range(len(shape))):
         labels = (x - y) % (factor * shape[i]) // factor
-        np.putmask(labels, comp_labels < 0, -1)
```
Comment (jbrockmendel) [inline pandas/core/sorting.py:221]: np.putmask is significantly more performatn

File: pandas/core/series.py [hunk 9]
```diff
@@ -2662,8 +2657,7 @@ def append(self, to_append, ignore_index=False, verify_integrity=False):
         from pandas.core.reshape.concat import concat
 
         if isinstance(to_append, (list, tuple)):
-            to_concat = [self]
-            to_concat.extend(to_append)
```
Comment (jbrockmendel) [inline pandas/core/series.py:2666]: i think this is more obvious than the tuple expansion

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -636,9 +636,7 @@ def _get_setitem_indexer(self, key):
         except TypeError as e:
 
             # invalid indexer type vs 'other' indexing errors
-            if "cannot do" in str(e):
-                raise
-            elif "unhashable type" in str(e):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:641]: coverage

File: pandas/core/generic.py [hunk 11]
```diff
@@ -8876,15 +8868,12 @@ def _align_frame(
...
         # if DatetimeIndex have different tz, convert to UTC
-        if is_datetime64tz_dtype(left.index.dtype):
-            if left.index.tz != right.index.tz:
-                if join_index is not None:
-                    # GH#33671 ensure we don't change the index on
-                    #  our original Series (NB: by default deep=False)
-                    left = left.copy()
-                    right = right.copy()
-                    left.index = join_index
-                    right.index = join_index
```
Comment (jbrockmendel) [inline pandas/core/generic.py:8887]: IIRC one of these two usages doesnt have test coverage

Comment (MarcoGorelli) [reply pandas/core/generic.py:8887]: true, but that's not affected by this PR - one of the two calls to `_set_join_index` will still be uncovered (as before) and the other one won't

File: pandas/core/sorting.py (hunk 8 above)
Comment (MarcoGorelli) [reply pandas/core/sorting.py:221]: isn't this what you'd requested in https://github.com/pandas-dev/pandas/pull/37586#discussion_r519237490 ? Have reverted anyway

File: pandas/core/algorithms.py [hunk 12]
```diff
@@ -108,11 +108,13 @@ def _ensure_data(
         values = extract_array(values, extract_numpy=True)
 
     # we check some simple dtypes first
-    if is_object_dtype(dtype):
-        return ensure_object(np.asarray(values)), np.dtype("object")
-    elif is_object_dtype(values) and dtype is None:
+    if (
+        (is_object_dtype(dtype))
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:112]: extra parens?

File: pandas/core/algorithms.py [hunk 13]
```diff
@@ -108,11 +108,13 @@ def _ensure_data(
...
+        or (not is_object_dtype(dtype))
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:113]: the or makes this unncessary doesnt it?  also extra parens (same below)

File: pandas/core/base.py [hunk 14]
```diff
@@ -611,8 +610,8 @@ def to_numpy(self, dtype=None, copy=False, na_value=lib.no_default, **kwargs):
         # TODO(GH-24345): Avoid potential double copy
         if copy or na_value is not lib.no_default:
             result = result.copy()
-            if na_value is not lib.no_default:
-                result[self.isna()] = na_value
```
Comment (jbrockmendel) [inline pandas/core/base.py:615]: why?  its small but we potentially avoid a duplicate check

File: pandas/core/common.py [hunk 15]
```diff
@@ -107,8 +107,10 @@ def is_bool_indexer(key: Any) -> bool:
             key = np.asarray(key)
 
             if not lib.is_bool_array(key):
-                na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
                 if lib.infer_dtype(key) == "boolean" and isna(key).any():
+                    na_msg = (
+                        "Cannot mask with non-boolean array containing NA / NaN values"
+                    )
```
Comment (jbrockmendel) [inline pandas/core/common.py:113]: why is this an improvement?

File: pandas/core/construction.py [hunk 16]
```diff
@@ -567,9 +566,13 @@ def _try_cast(arr, dtype: Optional[DtypeObj], copy: bool, raise_cast_failure: bo
         Otherwise an object array is returned.
     """
     # perf shortcut as this is the most common case
-    if isinstance(arr, np.ndarray):
-        if maybe_castable(arr) and not copy and dtype is None:
-            return arr
+    if (
+        isinstance(arr, np.ndarray)
+        and maybe_castable(arr)
```
Comment (jbrockmendel) [inline pandas/core/construction.py:571]: we should change maybe_castable to take arr.dtype directly

File: pandas/core/generic.py [hunk 17]
```diff
@@ -8849,17 +8839,19 @@ def _align_frame(
...
-                    other.columns, how=join, level=level, return_indexers=True
-                )
+        if (
+            (axis is None or axis == 1)
+            and not is_series
+            and not self.columns.equals(other.columns)
+        ):
+            join_columns, clidx, cridx = self.columns.join(
+                other.columns, how=join, level=level, return_indexers=True
+            )
```
Comment (jbrockmendel) [inline pandas/core/generic.py:8854]: i think could de-duplicate these be using `self.axes[axis]` and `other.axes[axis]`

File: pandas/core/generic.py [hunk 18]
```diff
@@ -11403,6 +11388,18 @@ def last_valid_index(self):
         return self._find_valid_index("last")
 
 
+def _set_join_index(
+    left: FrameOrSeries, right: FrameOrSeries, join_index: Index
+) -> Tuple[FrameOrSeries, FrameOrSeries]:
```
Comment (jbrockmendel) [inline pandas/core/generic.py:11393]: do we know that type(left) == type(right)?  if so, should this be FrameOrSeriesT?

Comment (MarcoGorelli) [reply pandas/core/generic.py:11393]: What's `FrameOrSeriesT`? I just tried searching for it in the codebase but couldn't find it

**Comment by jreback** [comment]
@MarcoGorelli can you rebase and see if you can get passing

**Comment by jreback** [comment]
ok this looks fine, but a ci /checks is failing. ping on greenish

File: pandas/core/algorithms.py [hunk 19]
```diff
@@ -114,11 +114,8 @@ def _ensure_data(
         values = extract_array(values, extract_numpy=True)
 
     # we check some simple dtypes first
-    if is_object_dtype(dtype):
+    if is_object_dtype(dtype) or (is_object_dtype(values) and (dtype is None)):
```
Comment (jreback) [inline pandas/core/algorithms.py:117]: can you remove the parens around dtype is None as confusing (and actually put around the entire expression and format)

**Review by jreback — CHANGES_REQUESTED** [review]
comment + merge master. master is moving, so ping on green when ready.

File: pandas/core/generic.py (hunk 1 above)
Comment (jreback) [reply pandas/core/generic.py:5500]: I am not sure about this change on L5509. pls revert and can do in a separate PR when checking perf. This causes recursion (which might be fine, though need to confirm)

File: pandas/core/generic.py [hunk 20]
```diff
@@ -11106,6 +11091,22 @@ def last_valid_index(self):
         return self._find_valid_index("last")
 
 
+def _set_join_index(
```
Comment (jreback) [inline pandas/core/generic.py:11094]: this should not be here but in pandas/core/reshape/merge.py or appropriate

**Review by jreback — CHANGES_REQUESTED** [review]
2 changes to revert. these types of PRs are very hard to review. in the future pls split out the non-controversial changes otherwise these linger forever.

ping on green when reverted

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli this is a nice cleanup. in the future pls do these types of changes in smaller batches.

**Pull request merged**